### PR TITLE
Chat page and create chat page

### LINKED ISSUE #231: Create a database table to persist the chat room
Part of the user story _As a user, I want to be able to create chat rooms so I can communicate with a group of users._

### LINKED ISSUE #232: Create a database method to add a chat room

### LINKED ISSUE #233: Create a POST /chatroom endpoint to add a chat room. This endpoint must be protected

### LINKED ISSUE #234: ⚛️  Create a page to add a chat room

### LINKED ISSUE #235: Create a database method adapter to read and filter chat rooms from the database

### LINKED ISSUE #236: Create a GET /chat/room endpoint to get the chat rooms
With the necessary parameters
This endpoint must be protected

### LINKED ISSUE #237: ⚛️  Create a page to list all the chat rooms

### LINKED ISSUE #244: ⚛️  Create a page to add password and name to a chat room

### PR DESCRIPTION
**Pull request by fizcamposancos**
close #231
close #232
close #233
close #234
close #235 
close #236
close #237
close #244
Please, let me know if you want me to split this PR in several PR to each of the tasks closed
Design is not closed, so any suggestion will be wellcome

### CODE REVIEW

**Comment by Damrod** [comment]
> Conflicting files
nginx/default.conf
transcendence-app/package-lock.json
transcendence-app/package.json
transcendence-app/src/app.module.ts
webapp/src/index.tsx
webapp/src/pages/index.ts
webapp/src/shared/urls.ts


This is what happens when we work on a branch that was created from a second branch that was merged into main via squashing.

File: webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx
Comment (gbudau) [inline webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx:32]: praise: I like how you structure this page and it works perfectly with the back-end :+1:

File: webapp/src/shared/urls.ts
Comment (gbudau) [inline webapp/src/shared/urls.ts:12]: question: What do we think about renaming this endpoint to `chat/create` or `chatroom/create`?

**Review by gbudau — COMMENTED** [review]
Thank you for this feature! I like how you structured the code.

File: transcendence-app/src/app.module.ts
Comment (Damrod) [inline transcendence-app/src/app.module.ts:7]: Why do you change this name back to Chat instead of ChatModule?

**Comment by Damrod** [comment]
After an `npm install`, my package-lock.json looks very different (more up to date)

**Comment by Damrod** [comment]
The chatroom button to go back is now broken.
This appears to fix it
```
diff --git a/webapp/src/pages/ChatRoom/ChatRoom.tsx b/webapp/src/pages/ChatRoom/ChatRoom.tsx
--- a/webapp/src/pages/ChatRoom/ChatRoom.tsx	(revision a9050428c26142a3f994e29eb888355c4dd42ebb)
+++ b/webapp/src/pages/ChatRoom/ChatRoom.tsx	(date 1666547191158)
@@ -29,7 +29,7 @@
   return (
     <div className="chat-room">
       <div className="chat-room-header">
-        <Header icon={IconVariant.ARROW_BACK} onClick={goBack}>
+        <Header icon={IconVariant.ARROW_BACK} onClick={goBack()}>
           {chatRoomId}
         </Header>
       </div>
```

File: webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx
Comment (Damrod) [inline webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx:152]: As of now users can write in private chats where they don't input the password. Is this something that is supposed to work?

File: db/setup/schema.sql
Comment (fizcamposancos) [reply db/setup/schema.sql:35]: Yes, looks good to me. Chats mean chatrooms in this table. I will do it

File: transcendence-app/package-lock.json
Comment (fizcamposancos) [reply transcendence-app/package-lock.json:4]: Yes, I had to change my computer to one with an old version of npm in a virtual machine and something happened. I will try to fix it

File: transcendence-app/src/chat/decorators/chat.decorator.ts
Comment (fizcamposancos) [reply transcendence-app/src/chat/decorators/chat.decorator.ts:3]: You are right, I will do it

File: webapp/src/shared/urls.ts
Comment (fizcamposancos) [reply webapp/src/shared/urls.ts:12]: I would say `chatroom/create` or `chat/createchatroom`, but I don't mind naming it any other way

File: transcendence-app/src/app.module.ts
Comment (fizcamposancos) [reply transcendence-app/src/app.module.ts:7]: Because I thought we were only going to have a ChatModule to manage everything, connections, joining and leaving chatrooms, ... In slack there is a thread about this, structure of chat modules of the backend. I will name it accordingly

File: webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx
Comment (fizcamposancos) [reply webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx:152]: No, not yet. That something left to do, I'm going to check if there is a ticket about it and, if not, I will create

File: transcendence-app/src/chat/chat.controller.ts
Comment (fizcamposancos) [reply transcendence-app/src/chat/chat.controller.ts:31]: You are right, it suits better with chatroom. I will change it

File: .env
Comment (gbudau) [inline .env:1]: suggestion: Please add this file to `.gitignore`

File: transcendence-app/src/chat/chat.controller.ts
Comment (gbudau) [inline transcendence-app/src/chat/chat.controller.ts:29]: suggestion: To use the same controller for the chatrooms and 1:1 chats, as we agreed in the last meeting, we could add the `'chat'` prefix to this controller and the `'room'` prefix to the POST and GET routes for the chat rooms.

```
@Controller('chat')
...
@Post('room')
...
@Get('room')
```
The routes to create a chat becomes `POST /chat/room` and to get all the chats `GET /chat/room`

**Review by gbudau — APPROVED** [review]
Thank you for this feature! LGTM :rocket:

File: transcendence-app/src/chat/chat.controller.ts
Comment (fizcamposancos) [reply transcendence-app/src/chat/chat.controller.ts:29]: Ok, I'll do it

File: transcendence-app/src/chat/infrastructure/db/postgres/chatPostgres.repository.ts
Comment (Damrod) [inline transcendence-app/src/chat/infrastructure/db/postgres/chatPostgres.repository.ts:1]: shouldn't the name of this file be chat.postgres.repository.ts?

File: transcendence-app/src/chat/chat.controller.ts
Comment (vdedios) [inline transcendence-app/src/chat/chat.controller.ts:44]: Following what @gbudau  suggested, and since this module will be common for 1:1 and chatrooms, I'd update all this nomenclatures to  `createChatRoom` vs `createChat` etc.

File: transcendence-app/src/chat/chat.controller.ts
Comment (vdedios) [inline transcendence-app/src/chat/chat.controller.ts:59]: Same here

File: transcendence-app/src/chat/chat.domain.ts
Comment (vdedios) [inline transcendence-app/src/chat/chat.domain.ts:5]: Will Chat and ChatRoom have the same domain model? we could export in the future , both models, Chat and ChatRoom, so I'd rename this to `ChatRoom`

File: transcendence-app/src/chat/chat.service.ts
Comment (vdedios) [inline transcendence-app/src/chat/chat.service.ts:26]: `retreiveChatRooms` ?

File: transcendence-app/src/chat/chat.service.ts
Comment (vdedios) [inline transcendence-app/src/chat/chat.service.ts:41]: `addChat` ?

File: transcendence-app/src/chat/chat.service.ts
Comment (vdedios) [inline transcendence-app/src/chat/chat.service.ts:53]: `createChatRoom` ?

File: transcendence-app/src/chat/chat.service.ts
Comment (vdedios) [inline transcendence-app/src/chat/chat.service.ts:68]: Very nice :heart:

File: transcendence-app/src/chat/chat.service.ts
Comment (vdedios) [reply transcendence-app/src/chat/chat.service.ts:41]: shouldn't this be a private method? or do we plan to expose this? I'd say with `createChat` would be enough

File: transcendence-app/src/chat/dto/update-chat.dto.ts
Comment (vdedios) [inline transcendence-app/src/chat/dto/update-chat.dto.ts:18]: I think we may rethink a bit how we implement avatars. They are also going to be used by chat and it may make sense to create a module for this in the future

File: transcendence-app/src/chat/infrastructure/db/postgres/chatPostgres.repository.ts
Comment (vdedios) [inline transcendence-app/src/chat/infrastructure/db/postgres/chatPostgres.repository.ts:62]: not yours, but I wanted to point this out. Since this query is being used in many places now, we may move it to the generic base repository. That is the idea, DRY principle! (don't repeat yourself :smile)

cc @gbudau @Damrod

File: webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx
Comment (vdedios) [inline webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx:50]: could you elaborate this hack? :smile:

File: webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx
Comment (vdedios) [inline webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx:81]: leftovers?

File: webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx
Comment (vdedios) [inline webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx:97]: leftover?

File: webapp/src/shared/context/auth-context.tsx
Comment (vdedios) [inline webapp/src/shared/context/auth-context.tsx:29]: Why?

**Review by vdedios — APPROVED** [review]
Very nice and huge PR :smile: !!  Very nice hard and good work :ok_hand: 

I've made several cosmetic comments and others to hear your point of view but I think this is near to be merged!

**Comment by Damrod** [comment]
Please, before merging this into master via squashing, allow me to take any and all new commits, so that I have all of your latest changes, and no merge conflicts

File: transcendence-app/src/chat/chat.service.ts
Comment (Damrod) [reply transcendence-app/src/chat/chat.service.ts:26]: rather `retrieveChatRooms` without a typo

File: transcendence-app/src/chat/chat.service.ts
Comment (Damrod) [reply transcendence-app/src/chat/chat.service.ts:68]: to be honest, this seems like it's mostly duplicated code from user. Maybe both User and Chat could extend a common BaseService ?

Comment (vdedios) [reply transcendence-app/src/chat/chat.service.ts:68]: Agree, I'd say to just include this in a small set of common utilities

File: transcendence-app/src/chat/infrastructure/db/postgres/chatPostgres.repository.ts
Comment (gbudau) [reply transcendence-app/src/chat/infrastructure/db/postgres/chatPostgres.repository.ts:62]: We could create a new task for this to not block this PR :+1:

File: transcendence-app/src/chat/chat.service.ts
Comment (gbudau) [reply transcendence-app/src/chat/chat.service.ts:68]: We could create a new task for this to not block this PR :+1:

File: webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx
Comment (fizcamposancos) [reply webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx:81]: Yes, sorry

File: webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx
Comment (fizcamposancos) [reply webapp/src/pages/CreateChatRoomPage/CreateChatRoomPage.tsx:97]: Yes, sorry

File: webapp/src/shared/context/auth-context.tsx
Comment (fizcamposancos) [reply webapp/src/shared/context/auth-context.tsx:29]: More leftovers of some test I did when something wasn't working as expected

**Comment by fizcamposancos** [comment]
I would say its ready @Damrod , @gbudau , @vdedios . Let me know how to when to merge and how

**Pull request merged**
